ADR/0006-log-conventions.md

```diff
@@ -0,0 +1,112 @@
+# 6. Log Conventions
+
+Date started: 2022-11-11
+Date revised: 2023-03-08
+
+## Status
+
+Accepted
+
+## Context
+
+We need log conventions to make our offering easier to operate and maintain. These will also make it possible for us to create queries that cross components in our logs, which supports [STONE-201](https://issues.redhat.com/browse/STONE-201).
+
+## Decision: Log conventions
+
+In our controller logs, we will use structured log messages, partially formatted in JSON with key-value pairs as described below.
+
+```
+TIMESTAMP LOG_LEVEL LOGGER MESSAGE ADDITIONAL_JSON_DATA
+```
+
+For example:
+
+```
+2023-03-07T11:32:29.167Z    INFO    controllers.Component   updating devfile component name kubernetes-deploy ...   {"resource": "samburrai-tenant/go-net-http-hello-prqx", "kind": "Component"}
+2023-03-07T11:32:29.020Z    INFO    controllers.Component   API Resource changed: UPDATE {"namespace": "samburrai-tenant", "audit": "true", "resource": "go-net-http-hello-prqx", "kind": "Component", "action": "UPDATE"}
+```
+
+The meaning of **timestamp** and log level should be self evident. The **logger** name helps your team understand which part of your controller is emitting the log (usually `Log.WithName` from `zap`). The **message** is a human readable string describing the event being logged. The **json data** contains additional fields useful for searching.
+
+### 1. When did it happen?
+
+Encode timestamps in UTC/ISO-8601 format.
+
+### 2. What happened?
```

Comment:
In the second example above, `UPDATE` is recorded twice. Do we want duplication of information in the string and JSON or should they be mutually exclusive and _supportive_ of each other?

Reply:
Good call. Will revise.

Reply:
WDYT of 043c00b?

Reply:
It is better that it was but still hard to parse what should be included at a quick glance. Maybe something like this would simplify the representation.

```
## Decision: Log conventions

## Decision: Questions to answer in logs

### 1. When did it happen?

**Included in:** `TIMESTAMP`

[...]

### 2. What happened?

**Included in:** `MESSAGE`, `ADDITIONAL_JSON_DATA`

[...]

### 3. Where did it happen?

**Included in:** _none_

[...]
```

To summarize the required keys in `ADDITIONAL_JSON_DATA`, we can have a simplified representation in the `Consequences` section if it makes sense.

Reply:
Updated in 842ec54.

Reply:
Cool, thanks.

Reply:
For the summarized ADDITIONAL_JSON_DATA, here's a jsonschema for it. Helpful? Or, overkill?

```json
{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "type": "object",
  "properties": {
    "action": {
      "description": "Indicates what action the log event is responding to",
      "type": "string",
      "enum": ["VIEW", "ADD", "UPDATE", "DELETE"]
    },
    "namespace": {
      "description": "Indicates the target namespace of the resource being reconciled",
      "type": "string"
    },
    "kind": {
      "description": "The kind of the resource being reconciled",
      "type": "string"
    },
    "resource": {
      "description": "The name of the resource being reconciled",
      "type": "string"
    },
    "source": {
      "description": "A url to the source code that is emitting this log",
      "type": "string"
    },
    "audit": {
      "description": "Set to true for specific types of audit logs required by SSML.PW.5.1.4",
      "type": "boolean"
    }
  },
  "required": [],
  "not" {
    "required": [
      "namespace_name",
      "container_name",
      "pod_name",
      "container_image",
      "pod_ip",
      "host",
      "hostname",
      "namespace_labels",
      "message",
      "level",
      "time"
    ]
  }
}
```

Reply:
That doesn't feel like overkill to me. It looks like a concise representation of the more detailed descriptions provided throughout!